### PR TITLE
[Enhancement] Add multiple results per Submission

### PR DESCRIPTION
**Pull request by anditurdiu**
### Checklist
- [x] I tested *all* changes and *all* related features with different users (student, tutor, instructor, admin) on the test server https://artemistest.ase.in.tum.de.
- [x] Server: I followed the [coding and design guidelines](https://artemis-platform.readthedocs.io/en/latest/dev/guidelines/server.html).
- [x] Server: I implemented the changes with a good performance and prevented too many database calls
- [x] Server: I documented the Java code using JavaDoc style.

### Motivation and Context
There is currently no supported way to allow for a second correction, see #2015. As a first step towards this goal, this PR aims to change the relation between the `Submission` and `Result` from: 
image

to:
image

**Note:** This PR only includes server changes.

### Description

- Change relation between Submission and Result from @OneToOne to @OneToMany.
- Add `results_order` column to `Result` to track the order of the results. Because of this change we also had to follow the `save` procedure described in https://artemis-platform.readthedocs.io/en/latest/dev/guidelines/database/#ordered
    - Add 0 to all existing results in their `results_order` column for backwards compatibility

- Perform the necessary code changes across the application to support the new relation
- Add custom serialization and deserialization tag for Jackson to serialize to a single result (latestResult) for the client, and to multiple results for the server. This is a temporary solution until we also change the relationship on the client. 

### Steps for Testing
There are no functional changes. The goal is to have everything working as usual. Therefore, we must test every aspect of the application that was affected by these changes. These include: 

- **Creating Text-, Modeling-, File Upload-, Quiz-, and Programming Exercises**
    - Add Example Submission with solution for modeling exercises **only**! (for text exercises there is a bug unrelated to this PR)

- **Creating Exam**
    - With all exercise types

- **Participating in Text-, Modeling-, File Upload-, Quiz-, and Programming Exercises**
    - Configure automatic programming assessment

- **Participating in Exam**

- **Conducting a Test Run**
    - Participate

- **Assessing Text-, Modeling-, File Upload-, Quiz-, and Programming Exercises**
    - Locking
    - Overriding
    - Submitting
    - Also for exam exercises

- **Complaining about the Assessment**
    - Complaint
    - More Feedback Request

- **Reviewing Complaint**
    - Accepting Complaint
    - Rejecting Complaint

### Test Coverage
There are no changes on the test coverage.

### CODE REVIEW

File: src/main/java/de/tum/in/www1/artemis/service/connectors/jenkins/JenkinsService.java [hunk 1]
```diff
@@ -322,7 +326,12 @@ public Result onBuildCompleted(ProgrammingExerciseParticipation participation, O
         final ProgrammingSubmission submission;
         submission = latestPendingSubmission.orElseGet(() -> createFallbackSubmission(participation, report));
         submission.setBuildFailed(result.getResultString().equals("No tests found"));
+
+        // todo wip, save currently necessary, because of relationship between result and submission
```
Comment (krusche) [inline src/main/java/de/tum/in/www1/artemis/service/connectors/jenkins/JenkinsService.java:330]: can we remove this?

File: src/main/java/de/tum/in/www1/artemis/service/connectors/jenkins/JenkinsService.java [hunk 2]
```diff
@@ -322,7 +326,12 @@ public Result onBuildCompleted(ProgrammingExerciseParticipation participation, O
...
+        resultRepository.save(result);
```
Comment (krusche) [inline src/main/java/de/tum/in/www1/artemis/service/connectors/jenkins/JenkinsService.java:331]:
```suggestion
        result = resultRepository.save(result);
```

File: src/main/java/de/tum/in/www1/artemis/service/SubmissionService.java [hunk 3]
```diff
@@ -287,10 +287,30 @@ public void hideDetails(Submission submission, User user) {
...
+        result = resultRepository.save(result);
         result.setSubmission(submission);
         submission.setResult(result);
-        result.setParticipation(submission.getParticipation());
+        submissionRepository.save(submission);
+        return result;
+    }
+
+    /**
+     * todo wip: currently used to fix issues with saving / assigning results to submissions
```
Comment (krusche) [inline src/main/java/de/tum/in/www1/artemis/service/SubmissionService.java:299]: can you add a better documentation here instead?
also make sure to write `TODO` Nicolas Ruscher or another name

File: src/main/java/de/tum/in/www1/artemis/service/connectors/bamboo/BambooService.java [hunk 4]
```diff
@@ -566,11 +570,17 @@ public Result onBuildCompleted(ProgrammingExerciseParticipation participation, O
             // Do not remove this save, otherwise Hibernate will throw an order column index null exception on saving the build logs
             programmingSubmission = programmingSubmissionRepository.save(programmingSubmission);
 
+            // todo wip, save currently necessary, because of relationship between result and submission
```
Comment (krusche) [inline src/main/java/de/tum/in/www1/artemis/service/connectors/bamboo/BambooService.java:573]: can we remove it and instead document why saving the result here is necessary?

**Comment by stefanwaldhauser** [comment]
I got the following error when trying to edit a modeling exercise that has an example submission attached:
image
https://artemistest.ase.in.tum.de/course-management/331/modeling-exercises
`detail":"Could not write JSON: failed to lazily initialize a collection of role: de.tum.in.www1.artemis.domain.Submission.results, could not initialize proxy - no Session; nested exception is com.fasterxml.jackson.databind.JsonMappingException: failed to lazily initialize a collection of role: de.tum.in.www1.artemis.domain.Submission.results, could not initialize proxy - no Session (through reference chain: de.tum.in.www1.artemis.domain.modeling.ModelingExercise[\"exampleSubmissions\"]->org.hibernate.collection.internal.PersistentSet[0]->de.tum.in.www1.artemis.domain.ExampleSubmission[\"submission\"]->de.tum.in.www1.artemis.domain.modeling.ModelingSubmission[\"result\"])","path":"/api/modeling-exercises/4837","message":"error.http.500"}`

File: src/main/java/de/tum/in/www1/artemis/service/connectors/bamboo/BambooService.java [hunk 5]
```diff
@@ -566,11 +570,17 @@ public Result onBuildCompleted(ProgrammingExerciseParticipation participation, O
...
+            result = resultRepository.save(result);
+
             var buildLogs = extractAndPrepareBuildLogs(buildResult, programmingSubmission);
             // Set the received logs in order to avoid duplicate entries (this removes existing logs)
+            result.setSubmission(programmingSubmission);
+            programmingSubmission.setResult(result);
+            programmingSubmission = programmingSubmissionRepository.save(programmingSubmission);
```
Comment (krusche) [inline src/main/java/de/tum/in/www1/artemis/service/connectors/bamboo/BambooService.java:580]: we now save the programmingSubmission twice (see two lines below), is this really necessary?

File: src/main/java/de/tum/in/www1/artemis/web/rest/ExerciseResource.java [hunk 6]
```diff
@@ -169,10 +173,10 @@ else if (authCheckService.isAtLeastTeachingAssistantForExercise(exercise, user))
         }
 
         // TODO CZ: load results of submissions eagerly to prevent additional database calls
-        List<ExampleSubmission> exampleSubmissions = this.exampleSubmissionRepository.findAllByExerciseId(exerciseId);
+        Set<ExampleSubmission> exampleSubmissions = this.exampleSubmissionRepository.findAllWithEagerResultByExerciseId(exerciseId);
```
Comment (krusche) [inline src/main/java/de/tum/in/www1/artemis/web/rest/ExerciseResource.java:176]: can we remove the above TODO now?

**Comment by FrankeLukas** [comment]
Got the following error upon reloading the quiz page after the timer ran out:
grafik

**Comment by stefanwaldhauser** [comment]
I can not create example submissions for text exercises
https://artemistest.ase.in.tum.de/course-management/331/text-exercises/4845/example-submissions/274
image




{"type":"https://www.jhipster.tech/problem/problem-with-message","title":"Internal Server Error","status":500,"detail":"Cannot invoke \"de.tum.in.www1.artemis.domain.Exercise.getCourseViaExerciseGroupOrCourseMember()\" because \"exercise\" is null","path":"/api/exercises/4845/example-submissions","message":"error.http.500"}

**Comment by FuchsDominik** [comment]
When clicking the red refresh button:
Screenshot 2020-12-01 at 11 10 22

**Comment by kloessst** [comment]
Probably unrelated:
image
After failing a build the build logs are only shown in the build log output. The result should also show Build Failed and make the Build Logs available

**Comment by anditurdiu** [comment]
Only for programming exam exercises: 
`{"type":"https://www.jhipster.tech/problem/problem-with-message","title":"Internal Server Error","status":500,"detail":"failed to lazily initialize a collection of role: de.tum.in.www1.artemis.domain.Result.feedbacks, could not initialize proxy - no Session","path":"/api/exercises/4846/programming-submission-without-assessment","message":"error.http.500"}`

**Review by fde312 — COMMENTED** [review]
image

**Comment by kloessst** [comment]
Also probably unrelated. This is for programming exercises with manual assessment:
1. Submit code that builds for a programming exercise in time
2. Submit code that fails to build in time

image
Course dashboard

image
Inside editor

It seems like the result of the failed build is ungraded, even though it was submitted in time.

**Comment by FrankeLukas** [comment]
The "Scores" view for the exams is not updated properly.

File: src/main/java/de/tum/in/www1/artemis/service/connectors/jenkins/JenkinsService.java (hunk 1 above)
Comment (anditurdiu) [reply src/main/java/de/tum/in/www1/artemis/service/connectors/jenkins/JenkinsService.java]: Removed in: ed4e902

File: src/main/java/de/tum/in/www1/artemis/service/connectors/jenkins/JenkinsService.java (hunk 2 above)
Comment (anditurdiu) [reply src/main/java/de/tum/in/www1/artemis/service/connectors/jenkins/JenkinsService.java]: Fixed in: ed4e902

File: src/main/java/de/tum/in/www1/artemis/service/connectors/bamboo/BambooService.java (hunk 4 above)
Comment (anditurdiu) [reply src/main/java/de/tum/in/www1/artemis/service/connectors/bamboo/BambooService.java]: adjusted in: ed4e902

File: src/main/java/de/tum/in/www1/artemis/service/connectors/bamboo/BambooService.java (hunk 5 above)
Comment (anditurdiu) [reply src/main/java/de/tum/in/www1/artemis/service/connectors/bamboo/BambooService.java:580]: Without the second save there would be an IndexColumnError null thrown for the build logs. We tried creating a BuildLogEntryRepository and first saving all BuildLogEntries, before establishing the relation but that did not fix the problem. This was the only solution @SimonEntholzer found. Maybe you can explain it a bit better?

File: src/main/java/de/tum/in/www1/artemis/service/SubmissionService.java (hunk 3 above)
Comment (anditurdiu) [reply src/main/java/de/tum/in/www1/artemis/service/SubmissionService.java]: adjusted in 00b31df355a6a714b5dc08d142c364095edabb35

**Comment by anditurdiu** [comment]
> I got the following error when trying to edit a modeling exercise that has an example submission attached:
> image
> https://artemistest.ase.in.tum.de/course-management/331/modeling-exercises
> `detail":"Could not write JSON: failed to lazily initialize a collection of role: de.tum.in.www1.artemis.domain.Submission.results, could not initialize proxy - no Session; nested exception is com.fasterxml.jackson.databind.JsonMappingException: failed to lazily initialize a collection of role: de.tum.in.www1.artemis.domain.Submission.results, could not initialize proxy - no Session (through reference chain: de.tum.in.www1.artemis.domain.modeling.ModelingExercise[\"exampleSubmissions\"]->org.hibernate.collection.internal.PersistentSet[0]->de.tum.in.www1.artemis.domain.ExampleSubmission[\"submission\"]->de.tum.in.www1.artemis.domain.modeling.ModelingSubmission[\"result\"])","path":"/api/modeling-exercises/4837","message":"error.http.500"}`
> image

Should be fixed after 11d67bf231a09b8c0f741f3512b1458c770813a3

File: src/main/java/de/tum/in/www1/artemis/web/rest/ExerciseResource.java (hunk 6 above)
Comment (anditurdiu) [reply src/main/java/de/tum/in/www1/artemis/web/rest/ExerciseResource.java:175]: Yes, removed the outdated todo in 0e13167fc5c6b10fe64381663e2fc069fcdc5357

**Comment by anditurdiu** [comment]
> The "Scores" view for the exams is not updated properly.

I analyzed this, this is actually not a bug. This happens because we filter for student participations only, and ignore instructors/administrator participations. This is because in reality users with instructor capabilities should not be able to participate in the exam. Instructors can only participate in the exam as part of test runs, the scores of which are excluded from the scores page. As you all participated with your admin accounts, your scores were disregarded in the scores page.

See:
image

I think long term we should not allow users to be registered to exams which have instructor or tutor access level in order to make sure that more people dont stumble on this behavior.

File: src/main/java/de/tum/in/www1/artemis/service/connectors/bamboo/BambooService.java (hunk 5 above)
Comment (fde312) [reply src/main/java/de/tum/in/www1/artemis/service/connectors/bamboo/BambooService.java:580]: You were right, this was not needed. There was still a issue with `null index column` for `buildLogEntries` which is fixed in `6444cd`.

**Comment by fde312** [comment]
@kloessst @FuchsDominik we should have now fixed the issues that you guys found during the testing session today. We can retest tomorrow all fixes in a testing session again :)

**Comment by fde312** [comment]
Issues found so for:
You cannot save a sample submission for modeling exercise (click save after everything was created)

question answers are not displayed even though correctly answered
image

File: src/main/java/de/tum/in/www1/artemis/config/SecurityConfiguration.java [hunk 7]
```diff
@@ -162,6 +162,7 @@ protected void configure(HttpSecurity http) throws Exception {
             .antMatchers("/time").permitAll()
         .and()
             .apply(securityConfigurerAdapter());
+
```
Comment (MaisiKoleni) [inline src/main/java/de/tum/in/www1/artemis/config/SecurityConfiguration.java:165]:
```suggestion
```

**Review by MaisiKoleni — COMMENTED** [review]
Works fine. Code looks good to me as far as I can see.

Only problem is that there are still some problems with the example submission, that the assessment cannot be saved correctly.

**Review by airelawaleria — COMMENTED** [review]
Tested in testing session. A bug with sample assessments where an error happens when updating the assessment(e.g. adding an explanation).

**Review by JohannesWeiss — COMMENTED** [review]
Tested in testing session, can confirm aforementioned issue.

**Review by marvinchimeka — COMMENTED** [review]
Tested on TS1. Everything worked as intended except the example submissions. During assessing the example submission there is an issue if you set field to 0 Points you cannot save your example submission. You get a translation not found exception. 
Bildschirmfoto 2020-12-02 um 22 40 01
 The only way to save your submission is when you put another value !=0 in the fields.

**Comment by fde312** [comment]
> Tested on TS1. Everything worked as intended except the example submissions. During assessing the example submission there is an issue if you set field to 0 Points you cannot save your example submission. You get a translation not found exception.
> Bildschirmfoto 2020-12-02 um 22 40 01
> The only way to save your submission is when you put another value !=0 in the fields.

Fixed in the last commit

**Comment by fde312** [comment]
> Tested in testing session. A bug with sample assessments where an error happens when updating the assessment(e.g. adding an explanation).

Fixed in `ac0e1f6`

**Review by fde312 — APPROVED** [review]
Tested it several times and all found issues were fixed!

**Review by fde312 — APPROVED** [review]

**Review by airelawaleria — APPROVED** [review]
All issues fixed. Approve

**Review by marvinchimeka — APPROVED** [review]
all issues resolved. Re-approve.

**Review by JohannesWeiss — APPROVED** [review]

**Review by valeriebucher — COMMENTED** [review]
works fine now!

**Pull request merged**